Fixed CubeTextureNode having incorrect roughness

DanielSturk opened the pull request:
Most devices support `EXT_shader_texture_lod`, so when doing the cube texture access in [envmap_physical_pars_fragment.glsl](https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/ShaderChunk/envmap_physical_pars_fragment.glsl.js#L17) we make use of `textureCubeLodEXT()`. Since `textureCubeLodEXT()` takes in the exact lod instead of a bias (`texCube()` uses a bias), they are not interchangeable.

[CubeTextureNode](https://github.com/mrdoob/three.js/blob/dev/examples/jsm/nodes/inputs/CubeTextureNode.js#L51) never tries to use `textureCubeLodEXT`, it always uses `texCube`. This effect is quite noticeable, thus causing a major inconsistency between the visual appearance of node materials and their counterparts. 

![12c1bef5e97d9c4b9cbe4a96b3a7a9e7](https://user-images.githubusercontent.com/2774286/60995411-793f8e00-a320-11e9-9dfd-903149d7dcbe.png)
![1c036541db66a4c2b6161db966e0b3a9](https://user-images.githubusercontent.com/2774286/60995416-7a70bb00-a320-11e9-97b7-c59481fefe6b.png)

DanielSturk committed:
Added demonstration of inconsistent reflection in webgl_materials_nodes.html

DanielSturk committed:
Fixed inconsistency in CubeTextureNode

DanielSturk commented:
In order to test this, only checkout the 1st of 2 commits. This will add additional options to the `nodes / reflect` demo in `webgl_materials_nodes.html`

By enabling/disabling the `node` option, you can see that the reflections are not the same in terms of roughness.

The next thing you should do is checkout all the commits (so far there's 2), and you will see this bug is fixed

DanielSturk commented:
@sunag

sunag commented:
Hi @DanielSturk 

It maybe is a bug, `texCubeBias`  should use `textureCubeLodEXT` if it is avaliable.
I would like to maintain a similar implementation like this today.:
https://github.com/mrdoob/three.js/blob/8d04711f57eb4c052f1cbe17cead4b0ae1ba06d0/examples/jsm/nodes/core/NodeBuilder.js#L79-L95

DanielSturk committed:
Reverted prev changes, simplified bug fix by changing use of TEXTURE_LOD_EXT

DanielSturk commented:
Thanks for the fast reply @sunag,
I didn't know you were already attempting to use `textureCubeLodEXT`. The reason yours didn't work is because `#define TEXTURE_LOD_EXT` will only be turned on when material.envMap is present. I can't imagine a reason why this would be necessary. I think it makes more sense for `TEXTURE_LOD_EXT` to only be coupled with `EXT_shader_texture_lod`, and not material.envMap, since that's what the name implies. Anyways, I reverted my changes and changed the behaviour of `#define TEXTURE_LOD_EXT`

sunag commented:
Niceee!
I also think `TEXTURE_LOD_EXT` should be used on custom materials too like this example.

Just to finish if possible add `color` and `side` params in standard material to stay with the same look.  This may be good for comparision purposes.
```
var standardMaterial = new THREE.MeshStandardMaterial( {
	color: nodeMaterial.color.value,
	side: defaultSide,
	envMap: cubemap,
	roughness: 0,
	metalness: 1
} );

mrdoob commented:
Test link:
https://raw.githack.com/DanielSturk/three.js/cubetexturenode-lod/examples/webgl_materials_nodes.html

mrdoob commented:
> I can't imagine a reason why this would be necessary. I think it makes more sense for `TEXTURE_LOD_EXT` to only be coupled with `EXT_shader_texture_lod`, and not material.envMap, since that's what the name implies. 

Shouldn't this line need to change too?
https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLProgram.js#L124

DanielSturk committed:
Made GL_EXT_shader_texture_lod extension declaration more concise

DanielSturk committed:
Removed whitespace

sunag commented:
> Shouldn't this line need to change too?
https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLProgram.js#L124

I am not sure we should change this part.

`generateExtensions()` is used to add defines to core materials based in your inputs and not the extension, right?, core materials do not use `material.extensions.shaderTextureLOD = true` as default? but use `material.envMap` like reference.
https://github.com/mrdoob/three.js/blob/d6ffa4d4aecf4ac37604b7be309141aa73f09303/src/renderers/webgl/WebGLProgram.js#L121-L124

sunag commented:
I replace `MeshLambertMaterial` to `MeshStandardMaterial` in `webgl_materials_cubemap` example and remove `parameters.envMap` of this line:
https://github.com/mrdoob/three.js/blob/d6ffa4d4aecf4ac37604b7be309141aa73f09303/src/renderers/webgl/WebGLProgram.js#L124

This crash the example with the error:
```
THREE.WebGLProgram: shader error:  0 35715 false gl.getProgramInfoLog invalid shaders
THREE.WebGLShader: gl.getShaderInfoLog() fragment
ERROR: 0:548: 'GL_EXT_shader_texture_lod' : extension is disabled
```

sunag commented:
Considering the observation of the @mrdoob is turn the code more concise and my tests avoid `ERROR: 0:548: 'GL_EXT_shader_texture_lod' : extension is disabled` bug of the PR, you can consider something like:
https://github.com/mrdoob/three.js/blob/b1a13ffbbcd40520da6f683f8c06f11a63031997/src/renderers/webgl/WebGLProgram.js#L528
To:
```javascript
( ( ! material.extensions || material.extensions.shaderTextureLOD ) || parameters.envMap ) && ( capabilities.isWebGL2 || extensions.get( 'EXT_shader_texture_lod' ) ) ? '#define TEXTURE_LOD_EXT' : '',
```
And keep this line how is it in core:
https://github.com/mrdoob/three.js/blob/b1a13ffbbcd40520da6f683f8c06f11a63031997/src/renderers/webgl/WebGLProgram.js#L124

DanielSturk committed:
revert code, included material.extensions in TEXTURE_LOD_EXT

DanielSturk commented:
Thank you @sunag, and sorry for the late reply...

I decided to replace my code with yours

I didn't realize the can of worms I was opening with this change. In the future, I think we should be enabling all the supported extensions (instead of enabling them based on the parameters), or at least auto-generate the material.extensions for the core materials based on parameters.

sunag commented:
> I didn't realize the can of worms I was opening with this change. In the future, I think we should be enabling all the supported extensions (instead of enabling them based on the parameters), or at least auto-generate the material.extensions for the core materials based on parameters.

Yeah. I think that many things like this can be naturally fixed and improved introducing `NodeMaterial` to `core`.

sunag commented:
@DanielSturk You can change this line again? This seems more simple and precise.
```javascript
( ( material.extensions ? material.extensions.shaderTextureLOD : false ) || parameters.envMap ) && ( capabilities.isWebGL2 || extensions.get( 'EXT_shader_texture_lod' ) ) ? '#define TEXTURE_LOD_EXT' : '',
```

DanielSturk committed:
Made `#define TEXTURE_LOD_EXT` more precise

sunag commented:
Thanks, @DanielSturk. @mrdoob I think that this already can be merged.

mrdoob commented:
Thanks!

DanielSturk commented:
Yay! Thanks @mrdoob @sunag
